hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java

```diff
@@ -142,9 +142,10 @@ public Set queryLocalMember(String mapName, Predicate predicate, IterationType i
             if (partitionIds.isEmpty()) {
                 return result;
             }
-        } catch (QueryResultSizeExceededException e) {
-            throw ExceptionUtil.rethrow(e);
         } catch (Throwable t) {
+            if (t.getCause() instanceof QueryResultSizeExceededException) {
```

Comment:
For me the biggest pain is that we have to catch the exception in so many places, not that the catch itself is not in the most elegant way right now.

So I would prefer to fix this in separate cleanup PR (maybe for 3.6). I'd like to de-duplicate big parts of the shared query logic and find a good solution to deal with this exception then.